### PR TITLE
add simple recipe for time-series search using TSPulse

### PR DESCRIPTION
**Pull request by t-sakai-kure**
Hi, I created a simple recipe for time-series search.
It uses the ShapesAll dataset from the UCR archive, reports the performance score, and visualizes the retrieved time series.

### CODE REVIEW

**Comment by wgifford** [comment]
@vijaye12 @t-sakai-kure Do you think that the `get_embedding` function can be externalized and made part of the library?

Are there other functions that should also be externalized?

**Comment by wgifford** [comment]
In the `get_embedding` function we seem to only pass `past_values` -- are there any use cases where the mask or other information is important?

**Comment by vijaye12** [comment]
> @vijaye12 @t-sakai-kure Do you think that the `get_embedding` function can be externalized and made part of the library?
> 
> Are there other functions that should also be externalized?

@t-sakai-kure - yes. can we move the get_embedding function here https://github.com/ibm-granite/granite-tsfm/blob/main/tsfm_public/models/tspulse/utils/helpers.py 

Also, can we make it more generic with extra mode option:

mode = "register" (default option) returning semantic embedding
mode = time (returning first n patch embeddings flattened)
mode = fft (returnning next n patch embeddings flattened)
mode = full (returning all embeddings time + fft + register)

**Comment by vijaye12** [comment]
> In the `get_embedding` function we seem to only pass `past_values` -- are there any use cases where the mask or other information is important?

@t-sakai-kure - kindly incorporate this change to make the get_embedding mor generic:

instead of passing mask_ratio = 0, pass mask_type = "user" [ remove mask_ratio=0 and add mask_type = "user"]
from 
`model = TSPulseForReconstruction.from_pretrained(model_path, num_input_channels=1, ,mask_ratio=0)


mask_type = "user" will do masking only based on "past_observed_mask" passed by the dataloader in batch. so - it will make mask_ratio 0 in normal cases by default unless dataloader/user wants to explicity pass this information..

in the model calling code, also pass past_observed_mask.  (see below)

```
_x = batch["past_values"]  # [B, T, C]

past_observed_mask = None
if "past_observed_mask" in batch:
  past_observed_mask = batch["past_observed_mask"]

embedding = model(x, past_observed_mask=past_observed_mask)
```

**Comment by vijaye12** [comment]
> past_observed_mask

@wgifford - I assume, any dataset class by default  in normal cases - will either return no "past_observed_mask" or set "past_observed_mask" to None or all 1s based on the past values shape..  

if na is there in the data, then it will be returned with 0s accordingly..

**Comment by t-sakai-kure** [comment]
Hi @wgifford @vijaye12, I moved the `get_embeddings` function to `tsfm_public/models/tspulse/utils/helpers.py`, and added support for additional arguments.
I also updated the notebook’s data loading logic to use the TSP trained on the training set for processing the test set. Additionally, I removed the normalization step in the `get_embeddings`, as TSP already handles normalization.

Another candidate for externalization is the `get_data` function. We could make `label_column` and `input_columns` configurable arguments. If `input_columns` is None, we can default to the standard UCR input columns, as currently done in the notebook.

File: tsfm_public/models/tspulse/utils/helpers.py [hunk 1]
```diff
@@ -199,3 +199,44 @@ def __getitem__(self, idx):
             **item,
             "past_observed_mask": past_observed_mask,
         }
+
+
+def get_embeddings(model, past_values, past_observed_mask=None, mode="register"):
```
Comment (wgifford) [inline tsfm_public/models/tspulse/utils/helpers.py:204]: Please add type hints and return type

**Comment by wgifford** [comment]
@t-sakai-kure one minor comment but I think this looks reasonable

**Comment by t-sakai-kure** [comment]
@vijaye12 Regarding the type hint for the model argument, is `TSPulseForReconstruction` sufficient as the type hint for the model arg? I noticed that some other model variants return `decoder_hidden_state`, but I am not sure how we can represent those with a single type hint. If you know one model class name, I will use it. If it is a bit complex, we can just write `TSPulseForReconstruction`  as the type hint and will update it later if necessary, or list several models using `List[model1, model2]`.

**Comment by vijaye12** [comment]
> past_observed_mask

its fine. we can also skip past_observed_mask as done currently..

**Comment by vijaye12** [comment]
> TSPulseForReconstruction

@t-sakai-kure  - can we use `model: Union[TSPulseForReconstruction, TSPulseForClassification]` ? 

also, can we add one more parameter in get_embedding to get embedding from either backbone or decoder. by default - it can be decoder which you are using currently ?

**Comment by vijaye12** [comment]
> > TSPulseForReconstruction

get_embedding:
      **component (str): component to use to get the embedding. Allowed values are backbone, decoder.**

File: tsfm_public/models/tspulse/utils/helpers.py (hunk 1 above)
Comment (t-sakai-kure) [reply tsfm_public/models/tspulse/utils/helpers.py]: @wgifford Thanks! The type hints are added: https://github.com/ibm-granite/granite-tsfm/pull/353/files#diff-58d11c8cdbbce05f3816db05139c46e46d3f5c2e07e2bd0f47af95a1d33ebce9R207

**Comment by t-sakai-kure** [comment]
Thanks @vijaye12. I added `Union[TSPulseForReconstruction, TSPulseForClassification]` as the type hint. Also, `component` parameter is added to `get_embedding`.

File: tsfm_public/models/tspulse/utils/helpers.py (hunk 1 above)
Comment (wgifford) [reply tsfm_public/models/tspulse/utils/helpers.py]: I think this is good. Later, if we decide to support more models, we can update as needed.

**Comment by wgifford** [comment]
> Another candidate for externalization is the `get_data` function. We could make `label_column` and `input_columns` configurable arguments. If `input_columns` is None, we can default to the standard UCR input columns, as currently done in the notebook.

We can hold off on this now -- we need to think about this in the context of the other utilities:
- timseries_preprocessor.get_datasets
- toolkit.data_handling.load_dataset
these should eventually be unified somehow, but I don't think it is critical at this point

**Comment by vijaye12** [comment]
@t-sakai-kure - I see - you have resolved most of the comments. can you mark the comments which are resolved as "resolved" ? so that we can see the pending and merge it ?

**Review by vijaye12 — APPROVED** [review]
